web/interceptor/access_control.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/quickfeed/quickfeed/web/auth"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+type (
+	role      int
+	roles     []role
+	requestID interface {
+		IDFor(string) uint64
+	}
+)
+
+const (
+	none role = iota
+	// user role implies that user attempts to access information about himself.
+	user
+	// group role implies that the user is a course student + a member of the given group.
+	group
+	// student role implies that the user is enrolled in the course with any role.
+	student
+	// teacher: user enrolled in the course with teacher status.
+	teacher
+	// courseAdmin: an admin user who is also enrolled into the course.
+	courseAdmin
+	// admin is the user with admin privileges.
+	admin
+)
+
+// If there are several roles that can call a method, a role with the least privilege must come first.
+// If method is not in the map, there is no restrictions to call it.
+var access = map[string]roles{
+	"GetUser":                {none},
+	"GetCourse":              {none},
+	"GetCourses":             {none},
+	"CreateEnrollment":       {user},
+	"UpdateCourseVisibility": {user},
+	"GetCoursesByUser":       {user},
+	// TODO(vera): needs a specific check: if request attempts to change admin role, user role is not sufficien.
+	"UpdateUser":              {user, admin},
+	"GetEnrollmentsByUser":    {user, admin},
+	"GetSubmissions":          {group, student, teacher, courseAdmin},
+	"GetGroupByUserAndCourse": {group, teacher},
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetAssignments":          {student, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"GetRepositories":         {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {teacher, courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
```

Comment:
Just to check; since we already have the `req` object here, would it be possible to do:
```go
request, ok := req.(requestID)
if !ok {
	// return early if the request does not implement the interface.
}
```
Or are there `req` types that don't implement the interface? If all expected `req` types were to implement the interface, we could always rely on the type assertion to pass. Then we could use the `request` in the remaining code, allowing us to call `request.IDFor(role)` without further type assertions. This would largely obviate the need for the `hasID()` function.

PS: I would perhaps change the `req interface{}` in the func signature to `request interface{}` and leaving `req` as is after the type assertion:
```go
req, ok := request.(requestID)
```

Reply:
There are requests like `GetUser`, `GetCourses` that send `Void` message which doesn't have any IDs. We can of course add a dummy `IDFor()` method for this message to avoid making exceptions.

Reply:
Yes. I think this makes good sense.